modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/frames/DtxRollback.java

```diff
@@ -62,6 +73,27 @@ protected void writeMethod(ByteBuf buf) {
     @Override
     public void handle(ChannelHandlerContext ctx, AmqpConnectionHandler connectionHandler) {
         int channelId = getChannel();
+        AmqpChannel channel = connectionHandler.getChannel(channelId);
+        XidImpl xid = new XidImpl(format, branchId.getBytes(), globalId.getBytes());
+
+        ctx.fireChannelRead((BlockingTask) () -> {
+            try {
+                channel.rollback(xid);
+            } catch (ValidationException e) {
+                LOGGER.debug("Validation error occurred while rolling back transaction", e);
+                ctx.writeAndFlush(new ChannelClose(getChannel(),
+                                                   ChannelException.PRECONDITION_FAILED,
+                                                   ShortString.parseString(e.getMessage()),
+                                                   CLASS_ID,
+                                                   METHOD_ID));
+            } catch (BrokerException e) {
+                LOGGER.error("Error occurred while rolling back transaction for xid " + xid, e);
+                ctx.writeAndFlush(new ConnectionClose(ConnectionException.INTERNAL_ERROR,
+                                                      ShortString.parseString(e.getMessage()),
+                                                      CLASS_ID,
+                                                      METHOD_ID));
+            }
+        });
         ctx.writeAndFlush(new DtxRollbackOk(channelId, XaResult.XA_OK.getValue()));
```

Comment:
Should we do the rollback in an async manner?

Reply:
Fixed!